Use refresh token to refresh access token after 1 h (resolved #91)

mjkrause opened the pull request:
The good news is that a user is not prompted to log in again after 1 h anymore. 

Yet, I have doubts about the way I've implemented it and would appreciate suggestions. Specifically, `_get_user_info_from_token` returns a 401 if access token isn't provided. That happens once user has been logged in for more than 1 h. I then set both, new refresh and access tokens in the User object instance. I'm not sure though why, following that, `current_user.access_token` doesn't return the refreshed access token?

The observation is that once 1 h has passed each new GET request immediately executes line  https://github.com/DataBiosphere/cgp-dashboard/blob/fix/mkrause/91-refresh-auth-token/app.py#L390, indicating that the refreshed access token wasn't set. In other words, I would expect a refreshed access token to be valid again for 60 min before it expires.

The code still contains several comments, logged in `dcc-dashboard/logs/error_uwsgiB.log` (after `sudo -i`) which helped me debugging it, and which I plan to remove.

An instance is running at https://mkrause1.ucsc-cgp-dev.org.

Commit:
Added refresh decorator to get_user_info and called it in authorization

In order to refresh the auth_token in a stale session (if user was
logged in) we use the login_manager's refresh handler.

Commit:
Reversed previous changes, forced immediate refresh

The access token expires only after 60 min. In order
to force using a refresh token (that's the part that
doesn't work currently), I set it to always true.

Commit:
WIP, useful change: set attribs for refresh and access token in User and a few comments an

Commit:
/callback now conforms more to 80-char rule, also easier to read

Commit:
Changes in get_user_info, passing user object around

Commit:
Reverting last commit

Method get_user_info(token=token, user=user) led to a ValueError.

Commit:
Merge remote-tracking branch 'origin/fix/mkrause/91-refresh-auth-token' into fix/mkrause/91-refresh-auth-token

Commit:
Added env var to app config to refresh cookie w/e request

Commit:
Added needs_refresh decorator to callback, reverted prev. commit

Commit:
Un-did some changes

Commit:
Removed user from /authorization

Commit:
Printing resp from new_google_access_token

Commit:
Setting resp status to 400 (true) to mimic expired access token

Commit:
Utilizing load_user method in several locations to get user object

Commit:
Method new_google_access_token now returns refresh token

Commit:
Corrected word

Commit:
Reverted response status code

Commit:
Logging (info) status code if 400 in get_user_info

This indicates that the access token is expired

Commit:
Set get_info to true and logged many variables

Commit:
Confirmed that log msgs are printed - reverting always true in get_info

Commit:
Status code from get_user_info if access token is expired is 401 (not 400)

Commit:
Removed input arg _user_ from get_user_info

Commit:
Inserted comment and removed an obsolete comment

Commit:
CHanged comments

jessebrennan commented on app.py:
```diff
@@ -374,12 +385,27 @@ def get_user_info(token=None):
+                            'Trying to refresh access token.'
+                            .format(resp.status_code))
+            access_token, refresh_token = new_google_access_token()
+            user = load_user(current_user.get_id())
```
You don't need to provide this argument to load user since it pulls the user's information from the User() class (which is a proxy for the cookie).

jessebrennan replied:
You don't need to provide this argument to load user since it pulls the user's information from the User() class (which is a proxy for the cookie).

jessebrennan commented on app.py:
```diff
@@ -374,12 +385,27 @@ def get_user_info(token=None):
+                            .format(resp.status_code))
+            access_token, refresh_token = new_google_access_token()
+            user = load_user(current_user.get_id())
+            setattr(user, 'access_token', access_token)
```
Using setattr here and below oversteps the use of the properties of the User object. I'm guessing this was intentional, but it shouldn't be done because accessing through the properties of the User objects automatically takes care of encryption and decryption of the tokens.

jessebrennan replied:
I was totally wrong about this, ignore.

jessebrennan commented on app.py:
```diff
@@ -481,8 +516,10 @@ def html_rend(name):
     Handle the templates differently depending
     on its name.
     """
+    # TODO: these variables are unused.
```
Good catch. Then they should be deleted

jessebrennan commented on app.py:
```diff
@@ -591,26 +663,49 @@ def callback():
         resp = google.get(Auth.USER_INFO)
+        app.logger.info("(MK comment) status code: {}".format(resp.status_code))
         if resp.status_code == 200:
+            user = load_user(current_user.get_id())
```
Again, load_user doesn't need the arg

jessebrennan commented on app.py:
```diff
@@ -591,26 +663,49 @@ def callback():
-            user.access_token = token['access_token']
+            #get_user_info(token=token, user=user)  # get new access token
+            for attr in 'refresh_token', 'access_token':
+                setattr(user, attr, token[attr])
```
Please don't use setattr here, it works around the encryption of the token.

jessebrennan replied:
I was wrong about this

jessebrennan commented on app.py:
```diff
@@ -591,26 +663,49 @@ def callback():
+                            'redirecting to boardwalk URL',
+                            request.path,
+                            user_data['email'])
+            return redirect(url_for('boardwalk'))
```
I think the redirect is fine to include in this PR (but others might disagree). What if it was just a separate commit

coverbeck replied:
Agreed. It is a separate commit already, and should be removed from here.

jessebrennan reviewed (changes requested):
Nice job working through to figure out how the token wasn't being refreshed. I'm wondering why the refresh token needs to be replaced, but I'll do a deeper review soon.

There's a handful of things that need to be changed. Also the logging you added could (should?) be removed.

coverbeck commented on app.py:
```diff
@@ -282,9 +284,14 @@ def parse_token():
+    
     authorization_header = request.headers.get("Authorization", None)
+    authorization_header2 = request.headers
+    app.logger.info('(MK comment) authorization header structure: {}'.format(authorization_header2))
```
We should not log user tokens.

coverbeck commented on app.py:
```diff
@@ -282,9 +284,14 @@ def parse_token():
+    app.logger.info('(MK comment) authorization header structure: {}'.format(authorization_header2))
     assert authorization_header is not None, "No Authorization header in the request"
     parts = authorization_header.split()
+    app.logger.info('(MK comment) parts[0]: {}, parts[1]: {}'.format(parts[0], parts[1]))
```
Should not log user tokens.

coverbeck commented on app.py:
```diff
@@ -303,8 +310,12 @@ def new_google_access_token():
+    app.logger.info('(MK comment) response from new_google_access_token: {}'.
+                    format(resp))
+    app.logger.info('(MK comment) refresh_token: {}'.format(refresh_token))
+    app.logger.info('(MK comment) resp[access_token]: {}'.format(resp['access_token']))    
```
Do not log user tokens.

coverbeck commented on app.py:
```diff
@@ -321,7 +332,7 @@ def make_request(url, headers):
         return e.message, e.code
 
-
+# What method calls check_session? (don't see any) What is its function?
```
This is a hold over from the old way of doing things. We should eventually remove it.

coverbeck commented on app.py:
```diff
@@ -374,12 +385,27 @@ def get_user_info(token=None):
+            access_token, refresh_token = new_google_access_token()
+            user = load_user(current_user.get_id())
+            setattr(user, 'access_token', access_token)
+            app.logger.info('(MK comment) set user.access_token: {}'
```
Don't log

coverbeck commented on app.py:
```diff
@@ -374,12 +385,27 @@ def get_user_info(token=None):
+                            .format(user.access_token))
+
+            setattr(user, 'refresh_token', refresh_token)
+            app.logger.info('(MK comment) set user.refresh_token: {}'
```
Don't log

coverbeck commented on app.py:
```diff
@@ -443,16 +469,24 @@ def authorization():
         # parsing succeeds if there is an auth header
         bearer, auth_token = parse_token()
-    except AssertionError:
+        app.logger.info('(MK comment) bearer: {}, auth_token: {}'.format(bearer, auth_token))
```
Don't log token

coverbeck commented on app.py:
```diff
@@ -443,16 +469,24 @@ def authorization():
-    except AssertionError:
+        app.logger.info('(MK comment) bearer: {}, auth_token: {}'.format(bearer, auth_token))
+    except AssertionError as e:
+        app.logger.info('(MK comment) Assertion error: ' + e.message)
```
Should be `.error` instead of `.info`

coverbeck commented on app.py:
```diff
@@ -543,43 +586,72 @@ def callback():
+                                 request.path, current_user.get_id())
             else:
-                app.logger.error('Request path %s. Access is denied for current user None', request.path)
+                app.logger.error('Request path %s. '
```
Why is this broken up to multiple arguments now? I don't know the behavior of the logger method.

coverbeck reviewed (changes requested):
Could you please remove the changes that are not directly related to the fix? It's hard to see what exactly changed.

In general, you should not have comments with your initials, and you should not log tokens. I tried to flag most of them, but I don't know that I got all of them. I would also revert the formatting changes you made. So that we can just see what changed for this bug fix.

It's great that you figured this out!

Thanks,

Charles

Commit:
Changed redirect back to index in method callback

Commit:
Changed get_user_info

* removed creation of user instance using the load_user method

* replaced user by current_user

* replace setattr assignment of value by simple assignement using =
to set access and refresh token in current_user

Commit:
Added REFRESH_URI (same as TOKEN_URI) to Auth class

Followed example of
https://github.com/requests/requests-oauthlib/blob/master/docs/examples/real_world_example_with_refresh.rst

Looks like syntactic sugar, but I suppose having it named explicitly
facilitates reading and understanding the code relevant to the token
refresh process.

Commit:
Logging the response status code after access token has been refreshed

Commit:
Removed method which accidentally was committed in previous commit

Commit:
Added parameter state to be included in _get_user_info_from_token

Commit:
New additions to improve refreshing the access token

* added new condition to get_google_auth
* get access token from session['oauth_token']
* set refreshed access token in session using refresh_token method
* added comment to log refreshed access token
* calling new_google_access_token without arguments (input or return)
  in get_user_info
